### ISSUE TITLE
Sonar reports 6 bugs (2 rated blocker, 4 rated critical)

### ISSUE THREAD

**Issue by rhowe** [dropwizard/dropwizard#4233]
Is it worth doing anything about this? https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&resolved=false&types=BUG

I've raised a few MRs to address sonar issues in the past where there were clear ways to both improve the code and also quieten the Sonar report, but the remaining group of 6 issues I'm either not sure about or seem like false positives.

I don't think having sonar grumbling is necessarily a bad thing, but if it's configured it would be nice if it had a clean report.

The issues:

- [dropwizard-auth ChainedAuthFilter](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdIbmpBosHuY9-P1w&resolved=false&types=BUG) can potentially throw an exception which can potentially be null. I'm not sure what the correct thing to do here is.
- [dropwizard-db TimeBoundHealthCheck](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXhd86P7aAnXFglb5z24&resolved=false&types=BUG) could potentially swallow an `InterruptedException`. This is possibly bad - should an extra `catch` clause be added? I'm not sure and threads scare me.
- [dropwizard-lifecycle ExecutorServiceBuilder](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdI01pBosHuY9-P9n&resolved=false&types=BUG) ignores the result of String.format, but intentionally so as it's using the call to check that the format string is valid. Sonar will stop complaining if it's wrapped in an explicit try-catch, but adding unnecessary code just to please Sonar isn't great. `@SuppressWarnings("ResultOfMethodCallIgnored")` doesn't seem to appease Sonar - is there anything else which will?
- [dropwizard-logging LoggingUtil](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdIKxpBosHuY9-Pub&resolved=false&types=BUG) - another potential swallowing of an `InterruptedException`. I have no idea what the right thing to do is here and messing with threads and logging frameworks doubly scares me.
- [dropwizard-logging ResilientSocketOutputStream](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdIKdpBosHuY9-PuY&resolved=false&types=BUG) creates a `Socket` but doesn't close it. This is entirely intentional as the method returns an active connection.
- [dropwizard-logging DropwizardUdpSocketAppender](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdILkpBosHuY9-Pug&resolved=false&types=BUG) creates a `DatagramSocket` but doesn't close it. This is entirely intentional as the method returns an active connection.

I'm not convinced any of these are real issues, although I admit to not being very confident about the thread-related ones.

What to do? Close this issue and let Sonar carry on grumbling to nobody in particular, add some suppression magic somewhere or make some code changes?

Mostly I'm raising this to say "I've looked at all the issues Sonar reported as major and I'm left with these. Is there anything I can do to clean stuff up further?"

**Comment by matthias-mueller** [comment]
An established pattern to handle a caught `InterruptedException` is this:

```java
try {
   ...
} catch (InterruptedException e) { // implicitly clears interrupted state
   // do something
   ...
   Thread.currentThread().interrupt(); // restore interrupted state
}
```

Here is a brief explanation: https://stackoverflow.com/a/4906814

**Comment by rhowe** [comment]
> An established pattern to not swallow a caught `InterruptedException` is this:

Yes, that is one of the things sonar suggests doing, but I don't know enough about the context in which the exception is swallowed to know whether it is intentional or not.

**Comment by matthias-mueller** [comment]
Since the interrupt flag was swallowed for the last 6 years (if sonar is right) the suggested solution should be fine. It will achieve the same thing as before but more safely. I assume the only time it is called is upon shutdown of the application or the JVM.

> Also does calling interrupt() on the current thread cause another InterruptedException to be thrown?

No. Restoring the interrupt state on the current thread won't (at least not directly) cause another `InterruptException`. There are two ways to deal with an `InterruptException`:
1) Abort your work and rethrow it (if you can)
2) Abort your work and restore interrupt state (without throwing it) - which seems to be your case

More thorough answers than mine seem to distinguish 3 cases (https://github.com/code-review-checklists/java-concurrency#thread-interruption-and-future-cancellation, https://stackoverflow.com/a/3976527), but so far I have only encountered the first two.

**Comment by rhowe** [comment]
>     2. Abort your work and restore interrupt state (without throwing it) - which seems to be your case

Alright, I've raised this for consideration:
https://github.com/dropwizard/dropwizard/pull/4239

**Comment by zUniQueX** [comment]
> dropwizard-lifecycle ExecutorServiceBuilder

The JavaDoc of `String.format` references `java.util.Formatter`. Maybe one could refactor the code to something like
`new Formatter().format(...).close()`, as this would be a `void` call then.

**Comment by zUniQueX** [comment]
> dropwizard-logging ResilientSocketOutputStream

Since the `Socket` is a local variable, it cannot be closed. Maybe it would be a good idea, to assign the `Socket` to a class field and close it in a call to an overridden `close` method. I don't know, it this makes Sonar happy, but the `Socket` handling would get improved anyway.

**Comment by zUniQueX** [comment]
> dropwizard-auth ChainedAuthFilter

The predefined `AuthFilter` implementations either set a new `SeciurityContext` through the `authenticate` method or throw a `WebApplicationException`. Therefore such a case cannot happen there.

Since calling the `authenticate` method to override the current `SecurityContext` of the `ContainerRequestContext` is not forced, one could provide a valid implementation, in which a user is authenticated as long as no exception is thrown.

Then the `firstException` should only be thrown, if it is not `null`.

**Comment by rhowe** [comment]
> > dropwizard-lifecycle ExecutorServiceBuilder
> 
> The JavaDoc of `String.format` references `java.util.Formatter`. Maybe one could refactor the code to something like
> `new Formatter().format(...).close()`, as this would be a `void` call then.

This looks viable. PR here: https://github.com/dropwizard/dropwizard/pull/4277

**Comment by zUniQueX** [comment]
> * [dropwizard-logging ResilientSocketOutputStream](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdIKdpBosHuY9-PuY&resolved=false&types=BUG) creates a `Socket` but doesn't close it. This is entirely intentional as the method returns an active connection.
> * [dropwizard-logging DropwizardUdpSocketAppender](https://sonarcloud.io/project/issues?id=dropwizard_dropwizard&open=AXUXdILkpBosHuY9-Pug&resolved=false&types=BUG) creates a `DatagramSocket` but doesn't close it. This is entirely intentional as the method returns an active connection.

@rhowe I've raised a PR for these: #4278

**Comment by zUniQueX** [comment]
> > dropwizard-auth ChainedAuthFilter
> 
> The predefined `AuthFilter` implementations either set a new `SeciurityContext` through the `authenticate` method or throw a `WebApplicationException`. Therefore such a case cannot happen there.
> 
> Since calling the `authenticate` method to override the current `SecurityContext` of the `ContainerRequestContext` is not forced, one could provide a valid implementation, in which a user is authenticated as long as no exception is thrown.
> 
> Then the `firstException` should only be thrown, if it is not `null`.

@joschi what do you think about this? Should a `null` check happen here or should a new `WebApplicationException` with status 401 be thrown?
